Sources/Plasma/CoreLib/hsSystemInfo.cpp

```diff
@@ -0,0 +1,333 @@
+                        s.Close();
+                        return system;
+                    }
+                }
+            }
+            s.Close();
+        }
+    }
+
+    return system;
+}
+
+ST::string hsSystemInfo::GetOperatingSystem()
+{
+    ST::string system;
+
+#if defined(HS_BUILD_FOR_WIN32)
+    const RTL_OSVERSIONINFOEXW& info = hsGetWindowsVersion();
+
+    typedef const char* (CDECL * wine_get_version)();
+    typedef void (CDECL * wine_get_host_version)(const char** sysname, const char** release);
+
+    HMODULE ntdll = GetModuleHandleW(L"ntdll");
+    if (ntdll) {
+        wine_get_version wine_target_version = (wine_get_version)GetProcAddress(ntdll, "wine_get_version");
+        wine_get_host_version wine_host_version = (wine_get_host_version)GetProcAddress(ntdll, "wine_get_host_version");
+        if (wine_target_version && wine_host_version) {
+            // WINE for some reason likes to pretend to be Windows Server. Debunk that crap
+            // because we really only want to get "Windows Server 2008 R2" if they are really
+            // on that OS, not because WINE is saying it's a Windows 7 server. Ugh.
+            ST::string windowsVersion = IGetWindowsVersion(info, false);
+
+            // Ideally, we would try to parse Z:\etc\os-release to get the Linux distro name,
+            // but that seems to crash WINE, unfortunately. So, we'll just naively use whatever
+            // WINE returns (the kernel name and version).
+            const char* sysname;
+            const char* release;
+            wine_host_version(&sysname, &release);
+            const char* wineVersion = wine_target_version();
+
+            system = ST::format("{} [Emulated on {} ({}) by WINE {}]", windowsVersion, sysname, release, wineVersion);
+        }
+    }
+    if (system.empty())
+        system = IGetWindowsVersion(info);
+#elif defined(HS_BUILD_FOR_APPLE)
+    CFDictionaryRef dict = _CFCopyServerVersionDictionary();
+    if (dict == nullptr)
+        dict = _CFCopySystemVersionDictionary();
+
+    if (dict != nullptr) {
+        CFStringRef name = (CFStringRef)CFDictionaryGetValue(dict, _kCFSystemVersionProductNameKey);
+        CFStringRef version = (CFStringRef)CFDictionaryGetValue(dict, _kCFSystemVersionProductVersionKey);
+        CFStringRef info = CFStringCreateWithFormat(nullptr, nullptr, CFSTR("%@ %@"), name, version);
+
+        CFIndex infoLen = CFStringGetLength(info);
+        CFIndex infoBufSz = 0;
+        CFStringGetBytes(info, CFRangeMake(0, infoLen), kCFStringEncodingUTF8, 0, false, nullptr, 0, &infoBufSz);
+        ST::char_buffer systemBuf;
+        systemBuf.allocate(infoBufSz);
+        CFStringGetBytes(info, CFRangeMake(0, infoLen), kCFStringEncodingUTF8, 0, false, (UInt8*)systemBuf.data(), infoLen, nullptr);
+        system = ST::string(systemBuf);
+
+        CFRelease(info);
+        CFRelease(version);
+        CFRelease(name);
+        CFRelease(dict);
+    }
```

Comment:
Does it make sense to split this out into a `IGetAppleVersion` function?

Reply:
Maybe. It sure is a lot of annoying boilerplate nonsense.